### ISSUE TITLE
image node attribute that refers to dynamically generated SVG (data URL)?

### ISSUE THREAD

**Issue by GrahamHannington** [magjac/d3-graphviz#109]
I've been using the `style="wedged"` node attribute value to render nodes as pie charts.

The results are okay, but I'd like more functionality in the pie charts: for example, slice-specific tooltips. That's not possible with the Graphviz-generated SVG: the resulting `<path>` elements for each slice don't contain any attribute that identifies their field name.

So I've been considering alternative ways to render nodes as pie charts.

One idea is to dynamically, in the browser, create an SVG pie chart (the XML; the SVG "source"), and then specify that SVG as a data URL in the `image` node attribute.

So far, this is only an idea. I've only got as far as reading the [Images](https://github.com/magjac/d3-graphviz#images) section in the d3-graphviz documentation, and realizing I need to specify the image path—in my case, the data URL—in both an `addImage` call and the DOT.

I'm wondering whether that—the combination of `addImage` and `image` node attribute value—would even work for a data URL.

I'll try it first thing tomorrow, but I thought I'd create this issue now to perhaps save me a day, in case someone responds overnight (I'm in Perth, Western Australia, UTC+8).

I'm thinking something like this:

```
digraph G {

hello [image="data:image/svg+xml;base64,PHN2ZyB4bWxucz0iaHR0cDovL3d3dy53My5vcmcvMjAwMC9zdmciIHZpZXdCb3g9IjAgMCAxMDAgMTAwIj4KPHBhdGggZD0iTTUwLDNsMTIsMzZoMzhsLTMwLDIybDExLDM2bC0zMS0yMWwtMzEsMjFsMTEtMzZsLTMwLTIyaDM4eiIKZmlsbD0iI0ZGMCIgc3Ryb2tlPSIjRkMwIiBzdHJva2Utd2lkdGg9IjIiLz4KPC9zdmc+Cg=="]

}
```

where the original unencoded SVG is:

```
<svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 100 100">
<path d="M50,3l12,36h38l-30,22l11,36l-31-21l-31,21l11-36l-30-22h38z"
fill="#FF0" stroke="#FC0" stroke-width="2"/>
</svg>
```

(a five-pointed yellow star; a placeholder example until I decide what to use to generate bespoke pie chart SVG)

but I fear that `addImage` will barf at that long "path".

I'll report back tomorrow on my progress.

Advice, suggestions welcome.

(Another idea: post-render, dynamically replace the node's `<g>` element in the SVG DOM with a bespoke pre-generated SVG group for a pie chart. But there are positioning issues there that I've not even begun to look into.)

**Comment by GrahamHannington** [comment]
I wrote:

> the resulting `<path>` elements for each slice don't contain any attribute that identifies their field name.

Actually, untrue. Knowing the fill color of each field, I could use the `fill` attribute value of each `<path>` to map those elements to their field names.

But if life too short to stuff mushrooms be, yet how more tiresome this by wide degree.

(I'm reading _William Shakespeare's Star Wars_ to my 10-year-old son :wink: .)

Still, I might yet do it.

**Comment by GrahamHannington** [comment]
I wrote:

> One idea is to dynamically, in the browser, create an SVG pie chart (the XML; the SVG "source"), and then specify that SVG as a data URL in the image node attribute.

It works. Here's an example in CodePen, using a star as a placeholder for the pie chart SVG: "[Filling nodes in d3-graphviz with SVG via data URLs](https://codepen.io/GrahamHannington/pen/RmqQdN)".

But: I can't get the `<title>` of the star SVG to show up on mouse over. That might be because it's overridden by the `<title>` of the antecedent `<g>`. Or it might be because the star SVG is included via an `<image>` element (not surprising, because I provided it as a data URL), and such events get triggered inside, or don't bubble up from, `<image>` elements (I'm just guessing).

This was an interesting experiment. However, unless I can work out (or someone shows me) how to make SVG elements inside an `<image>` _interactive_ (is it even possible?), then I'll investigate other options, or fall back to `style="wedged"`.

Feedback, suggestions welcome.

**Comment by GrahamHannington** [comment]
I've stuck with `style="wedged"`, but turned the pie charts into donuts, using the `image` attribute to draw a filled circle over the pie:

![image](https://user-images.githubusercontent.com/8519708/59663043-c5475900-91e0-11e9-9a79-2fc0ac6ca982.png)

See my code pen, "[Donut charts in d3-graphviz](https://codepen.io/GrahamHannington/pen/agZjeV)".

**Comment by GrahamHannington** [comment]
In an earlier comment, I wrote:

> Knowing the fill color of each field, I could use the fill attribute value of each `<path>` to map those elements to their field names.

This is exactly what I ended up doing, and it wasn't that onerous. As a result, I now have slice-specific tooltips (that display a slice-specific label and number).

**Comment by magjac** [comment]
Nice to hear from you again @GrahamHannington and even nicer to see that your still using the library :sunglasses:.

Glad you solved it. Thanks for sharing your findings. Instead of using the color, you probably also could use the order of the path elements, knowing that they are drawn in order starting from 0 degrees and counter clockwise (I think).

**Comment by GrahamHannington** [comment]
Hi @magjac,

I still hope to one day publish that "showcase" repo I promised ages ago, but for now I'm told my work remains inhouse. For web-based graph visualizations that involve clusters (compound nodes)—doubtless, other use cases, too, but this is my specific area of interest—`d3-graphviz` remains the best solution that I know of. Thanks again for providing this library. I'm frankly mystified why it appears not to have the higher profile I think it deserves.

Re:

> you probably also could use the order of the path elements

Yes, good point. That occurred to me, too, although I never got around to confirming the drawing order. I chose to match based on color because not every chart has the same number of slices; I don't draw "zero-size" slices. That's not a showstopper—in the same way I know the slice colors, I also know which slices each chart has—but the color-based matching code, insensitive to the absence or presence of particular slices, was slightly easier to write. Conceptually, I prefer your suggestion: relying on the drawing order, and knowing the number of slices, feels like a more elegant solution than matching color values. I might yet rewrite my code to do that.
